### PR TITLE
Support checking the entire class hierarchy when generating a hash

### PR DESCRIPTION
**Pull request by damoxc**
I've taken the changes from #337 and #338 and merged them into a single change, and added the minor changes requested from the original. Please let me know if any further changes are required.

### CODE REVIEW

File: src/DotLiquid.Tests/HashTests.cs [hunk 1]
```diff
@@ -189,4 +210,4 @@ public void TestMergeNestedDictionaries()
                 localVariables: hash);
         }
     }
-}
+}
```
Comment (microalps) [inline src/DotLiquid.Tests/HashTests.cs:213]: nit: Please keep newline at end of file

File: src/DotLiquid.Tests/HashTests.cs [hunk 2]
```diff
@@ -71,26 +80,38 @@ public void TestMapperCacheShouldCacheSeperateNamespaces()
...
+            Assert.AreEqual(
+                includeBaseClassProperties ? TestMiddleClassPropValue : null,
+                value[nameof(TestMiddleClass.TestMiddleClassProp)]);
 
             Assert.AreEqual(
                 includeBaseClassProperties ? TestBaseClassPropValue : null,
-                value[nameof(TestClass.TestBaseClassProp)]);
+                value[nameof(TestChildClass.TestBaseClassProp)]);
+
+            Assert.AreEqual(
```
Comment (microalps) [inline src/DotLiquid.Tests/HashTests.cs:108]: might want to keep this together with the other "ChildClass" property test. Would also help to comment
// Properties defined in TestChildClass should be exposed regardless of includeBaseClassProperties

and similarly for middle/base class tests

**Review by microalps — APPROVED** [review]

**Comment by microalps** [comment]
@daviburg please feel free to review. Looks good to me.

**Comment by damoxc** [comment]
@microalps thanks for the review, I've added your suggestions (I hope). Sorry about the EOF newline change, I didn't notice all the aggressive formatting Rider decided to do at one point, I tried to undo it all.

**Review by daviburg — APPROVED** [review]

**Comment by damoxc** [comment]
Thank you for merging so promptly!

**Pull request merged**